ouroboros-consensus/test-consensus/Test/ThreadNet/Network.hs

```diff
@@ -376,16 +376,21 @@ runThreadNetwork ThreadNetworkArgs
         loop s pInfo nr rs = do
           -- a registry solely for the resources of this specific node instance
           (again, finalChain) <- withRegistry $ \nodeRegistry -> do
-            nodeTestBtime <- cloneTestBlockchainTime
-              sharedTestBtime
-              nodeRegistry
-            let nodeBtime = testBlockchainTime nodeTestBtime
+            let nodeTestBtime = testBlockchainTimeClone sharedTestBtime nodeRegistry
+                nodeBtime     = testBlockchainTime nodeTestBtime
 
             -- change the node's key and prepare a delegation transaction if
             -- the node is restarting because it just rekeyed
             (pInfo', txs0) <- case (nr, mbRekeyM) of
               (NodeRekey, Just rekeyM) -> do
-                eno <- epochInfoEpoch epochInfo s
+                let ProtocolInfo{pInfoConfig} = pInfo
+                    k = protocolSecurityParam pInfoConfig
+                    -- the delegation will mature 2k slots after it is added to
+                    -- the chain
+                    --
+                    -- TODO we assume it's added in the current slot
+                    maturationSlot = s + 2 * SlotNo (maxRollbacks k)
+                eno <- epochInfoEpoch epochInfo maturationSlot
```

Comment:
TODO Revisit this.

This change eliminated the error I was seeing and has passed ~600 (?) test, but I haven't found a similar `2k` shift in the certificate validation code in the `cardano-ledger` package.

Reply:
I revisited this. I changed it, and I think it's right now. I introduced an oracle so that the node would know in which slot the next block will be forged.